node/adapter/datastore.py

```diff
@@ -314,14 +293,87 @@ def del_ip_pool(self, version, pool):
         assert version in ("v4", "v6")
         assert isinstance(pool, IPNetwork)
 
-        pools = self._get_ip_pools_with_keys(version)
+        pool_path = IP_POOL_PATH % {"version": version}
+        pools = self._get_path_with_keys(pool_path)
         try:
-            key = pools[pool.cidr]
+            key = pools[str(pool.cidr)]
             self.etcd_client.delete(key)
         except KeyError:
             # Re-raise with a better error message.
             raise KeyError("%s is not a configured IP pool." % pool)
 
+    def get_bgp_peer(self, version):
+        """
+        Get the configured BGP Peers
+
+        :param version: "v4" for IPv4, "v6" for IPv6
+        :return: List of netaddr.IPAddress IP addresses.
+        """
+        assert version in ("v4", "v6")
+        bgp_peer_path = BGP_PEER_PATH % {"version": version}
+        return map(IPAddress, self._get_path_with_keys(bgp_peer_path).keys())
+
+
+    def _get_path_with_keys(self, path):
+        """
+        Retrieve all the keys in a path and create a reverse dict
+        values -> keys
+
+        :param path: The path the the keys from.
+        :return: dict of {<values>: <etcd key>} for the peers.
+        """
+
+        try:
+            nodes = self.etcd_client.read(path).children
+        except KeyError:
+            # Path doesn't exist.  Interpret as no configured pools.
+            return {}
+        else:
+            values = {}
+            for child in nodes:
+                value = child.value
+                if value:
+                    values[value] = child.key
+            return values
+
+    def add_bgp_peer(self, version, ip):
+        """
+        Add a BGP Peer
```

Comment:
Mmmmmm, beer.

From: Spike Curtis [mailto:notifications@github.com]
Sent: Wednesday, April 15, 2015 3:05 PM
To: Metaswitch/calico-docker
Cc: Tom Denham
Subject: Re: [calico-docker] Add calicoctl commands for BGP Peers (and configurable AS) (#67)


In node/adapter/datastore.py<https://github.com/Metaswitch/calico-docker/pull/67#discussion_r28467848>:

> +        try:

> +            nodes = self.etcd_client.read(path).children

> +        except KeyError:

> +            # Path doesn't exist.  Interpret as no configured pools.

> +            return {}

> +        else:

> +            values = {}

> +            for child in nodes:

> +                value = child.value

> +                if value:

> +                    values[value] = child.key

> +            return values

> +

> +    def add_bgp_peer(self, version, ip):

> +        """

> +        Add a BGP Peer

Please comment that the method completes silently if the beer is already defined.

—
Reply to this email directly or view it on GitHub<https://github.com/Metaswitch/calico-docker/pull/67/files#r28467848>.